File: Exp 1/CRC.java
public class CRC {

    private static long  GetRemainder(String newString, String GenXString) {
        int r = GenXString.length() - 1;
        String subString = newString.substring(0, r);
        long mod = Long.parseLong(subString, 2);
        long divisor = Long.parseLong(GenXString, 2);
        for(int i = r; i < newString.length(); i++) {
            mod = mod * 2  + newString.charAt(i) - '0';
            if(mod >= Math.pow(2, r)) {
                mod ^= divisor;
            }
            else {
                mod ^= 0;
            }
        }
        return mod;
    }

    private static String GetSendString(String InfoString1, String GenXString) {
        int rr = GenXString.length() - 1;
        StringBuilder sbu = new StringBuilder(rr);
        for(int i = 0; i < rr; i++) sbu.append('0');
        String newString = InfoString1 + sbu.toString();
        long mod = GetRemainder(newString, GenXString);
        long tmp = Long.parseLong(InfoString1, 2) << rr;
        return Long.toBinaryString(tmp ^ mod);
    }

    public static void main(String[] argv) {
        String InfoString1 = "1101011111";
        String GenXString = "10011";
        String SendString = GetSendString(InfoString1, GenXString);
        System.out.println(SendString);
        long cmp = GetRemainder(SendString, GenXString);
        System.out.println(cmp);
    }
}

File: Exp 1/CRC.cpp
#include<iostream>
#include<string>
#include<sstream>
#include<fstream>
#include<bitset>
#include<ctime>
#include<math.h>
using namespace std;

long long int f(string s)//translate a binary string to long
{
	long long int a = 0;
	for (int i = 0; i < s.length(); i++)
	{
		a = a * 2 + (s[i] - '0');
	}
	return a;
}

string t(long long int a)//translate a long to binary 
{
	string s;
	while (a)
	{
		string temp = "1";
		temp[0] = a % 2 + '0';
		s = temp + s;
		a /= 2;
	}
	return s;
}

long long int GetRemainder(string newString, string GenXString)
{
	int r = GenXString.length() - 1;
	string subString = newString.substr(0, r);
	long long int mod = f(subString);
	long long int divisor = f(GenXString);
	for (int i = r; i < newString.length(); i++)
	{
		mod = mod * 2 + (newString[i] - '0');
		if (mod >= pow(2, r))
		{
			mod ^= divisor;
		}
		else
		{
			mod ^= 0;
		}
	}
	return mod;
}

string GetSendString(string InfoString1, string GenXString)
{
	int rr = GenXString.length() - 1;
	string newString = InfoString1;
	for (int i = 0; i < rr; i++)
	{
		newString = newString + "0";
	}
	long long mod = GetRemainder(newString, GenXString);
	long long tmp = f(newString);
	//cout << newString << endl;
	//cout << mod << endl;
	//cout << tmp << endl;
	cout << (mod^tmp) << endl;
	return t(tmp^mod);
}

void main()
{
	string InfoString1 = "1101011111";
	string GenXString = "10011";
	string SendString = GetSendString(InfoString1, GenXString);
	cout << SendString << endl;
	long long cmp = GetRemainder(SendString, GenXString);
	cout << cmp << endl;
	system("pause");
}
